echidna: 1.7.3 -> 2.0.2

arcz opened the pull request:
###### Description of changes
This fixes echidna that suffered from breakage of hevm. We are working on updating hevm to move forward with dependencies but it will take some time. Until then, both hevm and echidna need little patches to update aeson that changed API with the major update.

I tried to revert aeson to the previous version but that started a rabbit hole of incompatibilities.

Fixes #182500.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

arcz committed:
echidna: 1.7.3 -> 2.0.2

hellwolf commented:
what can we do to unblock this from being merged?

IvarWithoutBones commented on pkgs/tools/security/echidna/default.nix:
```diff
@@ -1,54 +1,68 @@
-  pname = "echidna";
-  version = "1.7.3";
+let
+  hevm = haskell.lib.unmarkBroken (
```
Shouldn't this be fixed in the `hevm` derivation instead? Might as well have it work in other derivations too.

IvarWithoutBones commented on pkgs/tools/security/echidna/default.nix:
```diff
@@ -1,54 +1,68 @@
+      platforms = lib.platforms.unix;
+      mainProgram = "echidna-test";
+    };
+in haskell.lib.compose.justStaticExecutables (
```
I think this should be moved back to `all-packages.nix`.

arcz commented:
@IvarWithoutBones what you pointed out makes total sense, I've made those changes. I originally started with another approach and didn't realize things can be moved around.

cdepillabout commented on pkgs/development/haskell-modules/configuration-nix.nix:
```diff
@@ -739,7 +739,7 @@ self: super: builtins.intersectAttrs super {
 
   # tests depend on a specific version of solc
-  hevm = dontCheck (doJailbreak super.hevm);
+  hevm = unmarkBroken (dontCheck (appendPatch ./patches/hevm-update-deps.patch super.hevm));
```
2. We almost never use `unmarkBroken`.  Instead, you'll need to remove `hevm` from the broken list in `pkgs/development/haskell-modules/configuration-hackage2nix/broken.yaml` and run `maintainers/scripts/haskell/regenerate-hackage-packages.sh`.
3. We try to never carry around `.patch` files in Nixpkgs because it bloats the repo.  If at all possible, we like to pull commits or PRs from upstream with `fetchpatch`.  There should be a bunch of examples of that in this file and `pkgs/development/haskell-modules/configuration-common.nix`.

arcz replied:
Fixed 2.

cdepillabout commented on pkgs/development/haskell-modules/patches/hevm-update-deps.patch:
```diff
@@ -0,0 +1,129 @@
+     transformers                      >= 0.5.6 && < 0.6,
+     tree-view                         >= 0.5 && < 0.6,
+     abstract-par                      >= 0.3.3 && < 0.4,
+-    aeson                             >= 1.5.6 && < 1.6,
```
4. Instead of patching these out, you could get away with just using `doJailbreak`.  That should require less maintenance in the long-run (but be slightly harder to debug when it ends up no longer working).

cdepillabout commented on pkgs/tools/security/echidna/default.nix:
```diff
@@ -1,5 +1,4 @@
 { lib
```
5. Is there some reason echidna is not on Hackage?  Could you upload it to hackage?  That way, you wouldn't have to maintain this file here, but it would get updated automatically with the rest of the Haskell packages in Nixpkgs.

arcz replied:
I have to speak with the rest of the echidna team, but I think we can do this.

arcz replied:
Is it okay if we merge this and I'll work on hackage for the next update?

cdepillabout replied:
Yeah, that sounds good, thanks!

cdepillabout reviewed (changes requested):
Thanks for sending this PR.

Couple pieces of feedback:

1. We generally ask all haskell-related PRs to go to the `haskell-updates` branch.  Could you rebase on there and change the base branch here on GitHub?

And then some more inline.

arcz commented:
@cdepillabout I rebased the PR. Regarding other comments, the situation with hevm is a bit problematic at the moment. I updated the dependencies in the upstream a while ago here https://github.com/dapphub/dapptools/pull/938 but the integration test suite there had issues after that. On top of that, recently, the development of hevm moved to https://github.com/ethereum/hevm with the deps updated but I'm waiting for the 0.50.0 release as the code is in development state right now. I think it's best to just keep those patches until 0.50.0 is published and I'll clean it all up when it lands.

cdepillabout commented:
@arcz So there is currently no commit in any upstream that makes these same compatibility changes?

arcz commented:
Right, I don't think there is a sensible commit to pick as a patch. I hand-crafted them.

cdepillabout commented:
@arcz Okay, in that case it sounds like we have no choice.  Could you make sure to put a good comment on them where you pull them in describing why they are necessary and when we expect to be able to get rid of them?

arcz commented:
@cdepillabout done

cdepillabout commented:
LGTM, thanks for working through this one!

If you're interested in making sure `echidna` and `hevm` don't break in the future, feel free to add yourselves as maintainers.  That way you'll get pinged if hackage updates ever cause echidna/hevm to stop compiling.
